```{r global_options, include=FALSE}
library(knitr)
opts_chunk$set(fig.align="center", fig.height=3, fig.width=4)
library(tidyverse) # ggplot2 library is contained within the tidyverse
theme_set(theme_bw(base_size = 12)) # set the default plot theme for the ggplot2
library(ggthemes) # enables colorblind-friendly color palettes
```
## Day 2: Data visualization
### In-class worksheet, solutions

**June 22nd, 2021**

## 1. Plotting x-y relationships

### 1.1 Correlations

We will work with the `iris` data set available in R. This data set gives the measurements in centimeters of the variables sepal length, sepal width, petal length and petal width for 50 flowers from each of 3 species of iris. The species are *Iris setosa*, *Iris versicolor*, and *Iris virginica*:

```{r}
# view the first several rows of the iris data set
head(iris)
```

Using ggplot, make a scatter plot of petal length vs. sepal length for the three species; the function you need for this is `geom_point()`. 

```{r}
# plot petal length vs sepal length, colored by species
ggplot(iris, aes(x = Sepal.Length, y = Petal.Length, color = Species)) + 
  geom_point()
```

Now, do the same plot but facet by species using `facet_wrap()`.

```{r} 
# plot petal length vs sepal length, faceted by species
ggplot(iris, aes(x = Sepal.Length, y = Petal.Length)) + 
  geom_point() + 
  facet_wrap(~Species)
```

### 1.2 Time series

The `omp` data set contains a subset of DNA microarray data measuring the differential expression of *E. coli* outer membrane proteins (omp) in nutrient-limited chemostatic cultures. In this particular experiment, the medium is glucose-limited. The `gene` column denotes one 8 genes that code for outer membrane proteins, `time_min` denotes the time point sampled in minutes, and `au` denotes the change in gene expression detected by the microarray chip (arbitrary units).

```{r}
# download the `omp` data set
omp <- read_csv("https://rachaelcox.github.io/classes/datasets/ecoli_omp_expression.csv")
head(omp)
```

Plot the expression of each gene over time using `geom_line()`, coloring each line by gene. Notice anything off about this plot?

```{r}
# plot the expression of each gene over time
ggplot(omp, aes(x = time_min, y = au, color = gene)) +
  geom_line()
```

Now, make the same plot but make the size of the lines thicker by specifying the `size` argument inside of `geom_line()`. Use `scale_color_colorblind()` to convert the legend colors to a colorblind-friendly palette. Use `xlab()` and `ylab` to give the figure pretty labels.

```{r}
# plot the expression of each gene over time & make it pretty
ggplot(omp, aes(x = time_min, y = au, color = gene)) +
  geom_line(size = 2) +
  scale_color_colorblind() +
  xlab("Time (min)") +
  ylab("Relative abundance (au)")
```

## 2. Plotting amounts

### 2.1 Bar plots

The `bacteria` data set contains data from tests of the presence of the bacterium *H. influenzae* in children with otitis media in the Northern Territory of Australia. We are interested in two columns of this data set: `presence` reports the presence (y) or absence (n) of the bacterium, `treatment` reports the treatment, which was placebo, drug, or drug+ (drug plus high adherence).

```{r}
# download the `bacteria` data set
bacteria <- read_csv("https://rachaelcox.github.io/classes/datasets/bacteria.csv")
head(bacteria)
```

Using `geom_bar()`, make a bar plot that shows the absolute number of cases with or without the bacterium, stacked on top of each other, for each treatment.

```{r }
ggplot(bacteria, aes(x = treatment, fill = presence)) + 
  geom_bar()
```

Now modify the plot so that bars representing the absolute number of cases with or without the bacterium are shown side-by-side. Use `scale_fill_brewer()` to change the plot colors. Hint: This requires the argument `position='dodge'` in `geom_bar()`.

```{r }
ggplot(bacteria, aes(x = treatment, fill = presence)) + 
  geom_bar(position = 'dodge') +
  scale_fill_brewer()
```

Now modify the plot so that bars represent the relative number of cases with or without the bacterium. What is the appropriate `position` option in `geom_bar()` to achieve this effect? Use `?geom_bar` to find out. Also, try setting an alternative color palette for `scale_fill_brewer()` to use by specifying the `type` and `palette` arguments (see `?scale_fill_brewer` for details).

```{r }
ggplot(bacteria, aes(x = treatment, fill = presence)) + 
  geom_bar(position = 'fill') +
  scale_fill_brewer(type = 'qual', palette = 'Set1')
```

### 2.2 Heat maps

The `dandelion` data set contains RNA-seq reads for a subset of genes differentially expressed in response to five conditions. Plot a heat map using `geom_tile` where each condition is on the x-axis and each gene (either `transcript_dandelion` or `loci_arabidopsis`) is on the y-axis. Fill the color of the heat map by the `log2_foldchange` value. Use `scale_fill_distiller()` to specify a continuous **diverging** color palette.

```{r}
# download the dandelion differential expression data set
dandelion <- read_csv("https://rachaelcox.github.io/classes/datasets/dandelion_diffexp_tidy.csv")
head(dandelion)
```

*(Note that I have told R Markdown to make a larger figure, by starting the code block with* `{r fig.height=6, fig.width=10}` *instead of* `{r}`*, because the default figure size is too narrow to show the resulting axes and map.)*

```{r fig.height=3.5, fig.width=7}
ggplot(dandelion, aes(x = condition, y = transcript_dandelion, fill = log2_foldchange)) +
  geom_tile() +
  scale_fill_distiller(type = 'div', 
                       palette = 'Spectral', 
                       guide = 'colourbar', 
                       direction = 1)
```

Make the same plot again, but rotate the text on the x-axis so they aren't mashing into each other. Try a different color palette this time.

```{r fig.height=4, fig.width=7}
ggplot(dandelion, aes(x = condition, y = transcript_dandelion, fill = log2_foldchange)) +
  geom_tile() +
  scale_fill_viridis_c() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  guides(title = 'log2(fc)')
```

## 3. Plotting distributions

### 3.1 Boxplots, violin plots

Using the `biopsy` data set, make side-by-side boxplots of `clump_thickness` for each `outcome`. The geom you need to use is `geom_boxplot()`.

```{r}
# download the biopsy data set
biopsy <- read_csv("https://rachaelcox.github.io/classes/datasets/biopsy.csv")
head(biopsy)
```

```{r}
ggplot(biopsy, aes(y = clump_thickness, x = outcome, fill = outcome)) + 
  geom_boxplot() +
  scale_fill_viridis_d(direction = -1, alpha = 0.75)
```

Now, make the same plot again, instead using `geom_violin()`. What do you notice?

```{r}
ggplot(biopsy, aes(y = clump_thickness, x = outcome, fill = outcome)) + 
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  scale_fill_viridis_d(direction = -1, alpha = 0.75, option = "E")
```

### 3.2 Histograms, density plots

Make a histogram plot of sepal lengths in the `iris` data set, using the default histogram settings. Then make two more such plots, with different bin widths. Use `geom_histogram()`. You can change the bin width by specifying the argument `binwidth =` or `bins = `. See `?geom_histogram` for more information.

```{r }
# default settings
ggplot(iris, aes(x = Sepal.Length)) + 
  geom_histogram()

# wider bins
ggplot(iris, aes(x = Sepal.Length)) + 
  geom_histogram(binwidth = 0.2)

# even wider bins
ggplot(iris, aes(x = Sepal.Length)) + 
  geom_histogram(binwidth = 0.4)

# divide all observations by a set # of bins
ggplot(iris, aes(x = Sepal.Length)) + 
  geom_histogram(bins = 10)
```

Instead of `geom_histogram()`, now use `geom_density()` and fill the area under the curves by species identity.

```{r}
ggplot(iris, aes(x = Sepal.Length, fill = Species)) + 
  geom_density()
```

Now make the areas under the curve partially transparent, so the overlap of the various distributions becomes clearly visible.

```{r}
ggplot(iris, aes(x = Sepal.Length, fill = Species)) + 
  geom_density(alpha = 0.7)
```

## 4. Layering geoms

**Bonus challenge:** For the `iris` data set, make a plot of the 2d distribution of petal length vs. sepal length, by making an x-y plot that shows the individual data points as well as contour lines indicating the density of points in a given spatial region.

```{r}
ggplot(iris, aes(x = Sepal.Length, y = Petal.Length, color = Species)) + 
  geom_point() + 
  geom_density2d()
```

Now instead of contour lines, add a fitted straight black line (not a curve, and no confidence band!) to each group of points. You'll need to check `?geom_smooth` to see which arguments you'll need to specify.

```{r}
ggplot(iris, aes(x = Sepal.Length, y = Petal.Length, color = Species)) + 
  geom_point() + 
  geom_smooth(
    aes(group = Species),  # make a line for each species
    method = lm,  # fit a linear model
    color = 'black',
    se = FALSE)  # no confidence interval
```

In this last example, because we are manually overriding the color of the lines, we need to set the group aesthetic to tell ggplot2 to draw a separate line for each species.

